aten/src/ATen/native/cuda/TensorFactories.cu

```diff
@@ -101,4 +104,199 @@ Tensor& randperm_out_cuda(Tensor& result, int64_t n, Generator* generator) {
   return result;
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ triangle ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+namespace {
+// f: the number of elements in the first row of the trapezoid.
+// x: the index of the target coordinates ordered by row and then column.
+// Assume x corresponding to coordinates (row, col) in the trapezoid, where
+// row and col start from 0, then we have:
+//     (f + f + row - 1) * row / 2 <= x < (f + f + row) * (row + 1) / 2.    [*]
+// Therefore, row is the maximum integer satisfying the following inequality:
+//              (row + 2f - 1)row <= 2x
+//         row^2 + (2f-1)row - 2x <= 0.
+// Based on the formula of root, we have:
+// a = 1
+// b = 2f - 1
+// c = -2x
+// Use the root on the right, intuitively because it is where the left-hand side
+// formular flips from negative to positive. Full proof can be derived from the
+// second inequality in [*].
+// row = floor((-b + sqrt(b^2 - 4c)) / 2)
+// col = x - (f + f + row - 1) * row / 2
+__device__
+inline void get_coordinate_in_tril_trapezoid(
+    int64_t f, int64_t x, int64_t & row, int64_t & col) {
+  f <<= 1;
+  auto b = f - 1;
+  auto c = - (x << 1);
+  row = (int64_t) ::floor((-b + ::sqrt(b * b - 4.0 * c))/2);
```

Comment:
Does this accumulate in double or float? Does this account for errors in precision?

Reply:
I'm a little worried about the numeric stability: can we run a test on some input with `b * b` much larger than `4 * c`?

Reply:
It accumulates in double. I will add some tests for large `b*b` and small `4 * c`.

Regarding the instability, I do see cases where the result was incorrectly rounded up when I was using float. Should I add back the check (if the # of elems up to `row` was larger than x), or do you know any other more stable way to calculate the row idx? I thought about binary search, but it would be much slower compared to directly calling `sqrt`.  

> would it be better to round the result of this to the nearest integer

Do you mean using the `__double2ll_rd` method?

Reply:
Double precision is good; double arithmetic on GPUs is slower than float arithmetic but it is better that this kernel is correct.

What exactly is unstable here? Is the problem taking the square root of a very big number?

> Do you mean using the __double2ll_rd method?

I'm not sure what _double2ll_rd is, but llround might work too: https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__DOUBLE.html#group__CUDA__MATH__DOUBLE_1g6e401c3a6f291b874fc95b8480bcad02

Reply:
When using single precision, I encountered an error while calculating indices for a 5k X 5k matrices. The result should be something like 4999.999998, but got rounded up to 5000 as float only has 7-8 digit precision.